### ISSUE TITLE
fail to recognize c++ keywords "const override"

### ISSUE THREAD

**Issue by shawn-xy** [universal-ctags/ctags#1180]
Hi guys,

First of all, thanks for developing the great tool! I use it everyday.

Now the issue: for the following code,

class B {
    public:
        virtual void foo() {
        }
        virtual void bar() const {
        }
        virtual void foobar() {
        }
};
class D : public B {
    public:
        virtual void foo() override {
        }
        virtual void bar() const override {
        }
        virtual void foobar() override {
        }
};

and comand:
$ ctags -R 
1. both B::foo and D::foo are recognized;
2. so is B::bar but not D::bar;
3. so is B::foobar but not D::foobar.

Seems that it fail on key word sequence “const override” and stopped working.

I run it on Mac and installed by brew. Here is the version info:
$ ctags --version
Universal Ctags 0.0.0(ffe4bc4), Copyright (C) 2015 Universal Ctags Team
Universal Ctags is derived from Exuberant Ctags.
Exuberant Ctags 5.8, Copyright (C) 1996-2009 Darren Hiebert
  Compiled: Nov 11 2016, 13:36:22
  URL: https://ctags.io/
  Optional compiled features: +wildcards, +regex, +option-directory, +coproc, +xpath

Best Regards,
Shawn

**Comment by masatake** [comment]
A bit simplified input:
```
class X {
public:
  virtual void m () const {
  }
};

class Y: X {
public:
  void m () const override {
  }
};

```

The second `m` is not picked.

**Comment by masatake** [comment]
Thank you for reportin.  @pragmaware, the maintainer of cxx parser may be in busy.
So I provide a temporary solution here:

```
index 58855a6..268e3e4 100644
--- a/parsers/cxx/cxx_parser_block.c
+++ b/parsers/cxx/cxx_parser_block.c
@@ -57,6 +57,7 @@ static bool cxxParserParseBlockHandleOpeningBracket(void)
                                                // new T { arg1, arg2, ... } (3)
                                                // Class::Class() : member { arg1, arg2, ... } { (4)
                                                cxxTokenTypeIs(g_cxx.pToken->pPrev,CXXTokenTypeIdentifier) &&
+                                               ( strcmp (vStringValue (g_cxx.pToken->pPrev->pszWord), "override")) &&
                                                (
                                                        (!g_cxx.pToken->pPrev->pPrev) ||
                                                        (cxxTokenTypeIsOneOf(

```


I got following output with the patch
```
[yamato@x201]~/var/ctags-github% cat const-override.cpp                                                                 
class Y {
  void m () const override {
  }
};
[yamato@x201]~/var/ctags-github% ./ctags --sort=no -o - const-override.cpp
Y	const-override.cpp	/^class Y {$/;"	c	file:
m	const-override.cpp	/^  void m () const override {$/;"	f	class:Y	typeref:typename:void	file:

```

Please, wait for a complete fix till the maintainer comes back.

**Comment by pragmaware** [comment]
@masatake : your fix is good. Applying it.
See #1182

**Comment by masatake** [comment]
@pragmaware, thank you.

As usual, CXX_DEBUGGING_ENABLED is really helpful to understand the parser.
When we got time, we should move to main part so that the other parsers can use the facility.

**Comment by pragmaware** [comment]
Closed via #1182

**Comment by shawn-xy** [comment]
Thank you guys!

The update works with "const override" now, but fail with some macros defined like:
`
#define OVERRIDE override
class Y : class X {
virtual void m() const OVERRIDE{}
}
`
My project happens to have macros like this.

I'm using a temporal fix locally:

`index ce79d6d..55b374e 100644
--- a/parsers/cxx/cxx_parser_block.c
+++ b/parsers/cxx/cxx_parser_block.c
@@ -25,6 +25,22 @@

 #include <string.h>

+static bool cxxTokenHasPrecedingParenthesisChain(CXXToken* pToken) {
+       bool ret = false;
+       CXXToken * pAux = pToken->pPrev;
+       while (pAux) {
+               if (cxxTokenTypeIsOneOf(pAux, CXXTokenTypeClosingBracket | CXXTokenTypeSemicolon)) {
+                       break;
+               }
+               if (cxxTokenTypeIs(pAux, CXXTokenTypeParenthesisChain) ) {
+                       ret = true;
+                       break;
+               }
+               pAux = pAux->pPrev;
+       }
+       return ret;
+}
+
 bool cxxParserParseBlockHandleOpeningBracket(void)
 {
        CXX_DEBUG_ENTER();
@@ -67,8 +83,7 @@ bool cxxParserParseBlockHandleOpeningBracket(void)
                                                                        CXXTokenTypeSingleColon | CXXTokenTypeComma
                                                        ))
                                                ) &&
-                                               // "override" is handled as identifier since it's a keyword only after function signatures
-                                               (strcmp(vStringValue(g_cxx.pToken->pPrev->pszWord),"override") != 0)
+                                               (!cxxTokenHasPrecedingParenthesisChain(g_cxx.pToken->pPrev))
                                        ) || (
                                                // type var[][][]..[] { ... }
                                                // (but not '[] { ... }' which is a parameterelss lambda)`


BR,
Shawn
